crates/sc-subspace-block-relay/src/types.rs

```diff
@@ -0,0 +1,105 @@
+//! Common type definitions.
+
+use codec::{Decode, Encode};
+use sc_network::RequestFailure;
+
+/// The protocol identifier.
+pub(crate) type RelayProtocol = u64;
+
+/// The overall protocol version, used in the exchanged
+/// messages.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Encode, Decode)]
+pub(crate) struct RelayVersion {
```

Comment:
https://github.com/subspace/subspace/pull/1937#discussion_r1327835988 continuing from there since I'm not able to add a comment through review

>No, this is exactly what I want to avoid (also mentioned in https://github.com/subspace/subspace/pull/1937#discussion_r1326550384), about creating global enum and breaking the protocol abstraction

Not sure I follow you. How is it breaking abstraction? The request and response defines themselves which protocol to use and how to handle them. In the current impl, we seem to be passing these version identifiers to detect them and we can completely remove them with self defined protocol messages, no?

Can you help me understand what I'm missing here ?

Reply:
My understanding matched Ved's. And I like the idea of full block download being its own protocol in another comment. Each protocol should only care about messages it can understand and can be each in their own module or even crate (depending on size). Then high-level multiplexer simply forwards messages to corresponding protocols for actual processing after simple matching.

Reply:
What we're suggesting is exactly the same: you have an enum that has inner implementation protocols. None of the protocols know or care about each other. As mentioned above they can even live in separate crates.

I'm not suggesting to make it monolithic, at least the way I understand it.

We'll start with full block download for ranges/old blokcs and compact blocks for individual blocks on gossip, we can add graphene or any other protocol later when we need it, but we need our infra to support it first.

Reply:
Assuming this is what you are still talking about: https://github.com/subspace/subspace/pull/1937#discussion_r1326907302

Say a new protocol needs to be added, not sure how that is possible without adding it in the global `ProtocolMessage` enum in your example. Pls explain what I am missing.

>  We'll start with full block download for ranges/old blokcs and compact blocks for individual blocks on gossip, we can add graphene or any other protocol later when we need it, but we need our infra to support it first.

Don't get what this means

Don't want to drag this, hope we can resolve this soon.

Reply:
> Say a new protocol needs to be added, not sure how that is possible without adding it in the global ProtocolMessage enum in your example

But that alongside with a small piece of code is the only thing that knows about all protocol. Protocols themselves are not aware of it. Something somewhere must combine it all together at some point, you can't really avoid that completely.

> Don't get what this means

This is an additional piece of logic that makes decision about which protocol to use for outgoing requests depending on what is being requested. Individual protocols do not know about this, but something on mutiplexer level must be able to make a decision which protocol to use: full block download or Compact blocks/Graphene.